File: linda-core/src/main/java/it/unibo/coordination/linda/core/events/TupleEvent.java
package it.unibo.coordination.linda.core.events;

import it.unibo.coordination.linda.core.Template;
import it.unibo.coordination.linda.core.Tuple;
import it.unibo.coordination.linda.core.TupleSpace;

import java.util.Objects;

public class TupleEvent<T extends Tuple, TT extends Template> extends TupleSpaceEvent<T, TT> {

    public enum Effect { WRITTEN, READ, TAKEN, ABSENT };

    private final Effect effect;
    private final boolean before;
    private final T tuple;
    private final TT template;

    private TupleEvent(TupleSpace<T, TT> tupleSpace, boolean before, Effect effect, T tuple, TT template) {
        super(tupleSpace);
        this.before = before;
        this.effect = Objects.requireNonNull(effect);
        this.tuple = tuple;
        this.template = template;
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> beforeWriting(TupleSpace<X, Y> tupleSpace, X tuple) {
        return new TupleEvent<>(tupleSpace, true, Effect.WRITTEN, tuple, null);
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> afterWriting(TupleSpace<X, Y> tupleSpace, X tuple) {
        return new TupleEvent<>(tupleSpace, false, Effect.WRITTEN, tuple, null);
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> beforeTaking(TupleSpace<X, Y> tupleSpace, X tuple) {
        return new TupleEvent<>(tupleSpace, true, Effect.TAKEN, tuple, null);
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> afterTaking(TupleSpace<X, Y> tupleSpace, X tuple) {
        return new TupleEvent<>(tupleSpace, false, Effect.TAKEN, tuple, null);
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> beforeReading(TupleSpace<X, Y> tupleSpace, X tuple) {
        return new TupleEvent<>(tupleSpace, true, Effect.READ, tuple, null);
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> afterReading(TupleSpace<X, Y> tupleSpace, X tuple) {
        return new TupleEvent<>(tupleSpace, false, Effect.READ, tuple, null);
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> beforeAbsent(TupleSpace<X, Y> tupleSpace, Y template) {
        return new TupleEvent<>(tupleSpace, true, Effect.ABSENT, null, template);
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> afterAbsent(TupleSpace<X, Y> tupleSpace, Y template) {
        return new TupleEvent<>(tupleSpace, false, Effect.ABSENT, null, template);
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> beforeAbsent(TupleSpace<X, Y> tupleSpace, Y template, X counterexample) {
        return new TupleEvent<>(tupleSpace, true, Effect.ABSENT, counterexample, template);
    }

    public static <X extends Tuple, Y extends Template> TupleEvent<X, Y> afterAbsent(TupleSpace<X, Y> tupleSpace, Y template, X counterexample) {
        return new TupleEvent<>(tupleSpace, false, Effect.ABSENT, counterexample, template);
    }

    public Effect getEffect() {
        return effect;
    }

    public boolean isBefore() {
        return before;
    }

    public boolean isAfter() {
        return !before;
    }

    public T getTuple() {
        return tuple;
    }

    public TT getTemplate() {
        return template;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        TupleEvent<?, ?> that = (TupleEvent<?, ?>) o;
        return before == that.before &&
                Objects.equals(effect, that.effect) &&
                Objects.equals(tuple, that.tuple) &&
                Objects.equals(template, that.template);
    }

    @Override
    public String toString() {
        return "TupleEvent{" +
                "tupleSpace=" + getTupleSpaceName() +
                ", effect='" + effect + '\'' +
                ", before=" + before +
                ", tuple=" + tuple +
                ", template=" + template +
                '}';
    }

    @Override
    public int hashCode() {
        return Objects.hash(effect, before, tuple, template);
    }
}

File: linda-core/src/main/kotlin/it/unibo/coordination/linda/core/events/TupleEvent.kt
package it.unibo.coordination.linda.core.events

import it.unibo.coordination.linda.core.Template
import it.unibo.coordination.linda.core.Tuple

sealed class TupleEvent<T : Tuple<T>, TT : Template<T>>(
        override val tupleSpaceName: String,
        open val isBefore: Boolean,
        val effect: Effect,
        open val tuple: T?,
        open val template: TT?
) : TupleSpaceEvent<T, TT> {

    enum class Effect {
        WRITING, READING, TAKING, MISSING
    }

    val isAfter: Boolean
        get() = !isBefore

    data class Writing<T : Tuple<T>, TT : Template<T>>(
            override val tupleSpaceName: String,
            override val isBefore: Boolean = true,
            override val tuple: T
    ) : TupleEvent<T, TT>(tupleSpaceName, isBefore, Effect.WRITING, tuple, null)

    data class Reading<T : Tuple<T>, TT : Template<T>>(
            override val tupleSpaceName: String,
            override val isBefore: Boolean = true,
            override val tuple: T
    ) : TupleEvent<T, TT>(tupleSpaceName, isBefore, Effect.READING, tuple, null)

    data class Taking<T : Tuple<T>, TT : Template<T>>(
            override val tupleSpaceName: String,
            override val isBefore: Boolean = true,
            override val tuple: T
    ) : TupleEvent<T, TT>(tupleSpaceName, isBefore, Effect.TAKING, tuple, null)

    data class Missing<T : Tuple<T>, TT : Template<T>>(
            override val tupleSpaceName: String,
            override val isBefore: Boolean = true,
            override val template: TT,
            override val tuple: T? = null
    ) : TupleEvent<T, TT>(tupleSpaceName, isBefore, Effect.MISSING, tuple, template) {

        val counterExample
            get() = tuple
    }

    companion object {

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> of(tupleSpaceName: String, before: Boolean, effect: Effect, tuple: X?, template: Y?): TupleEvent<X, Y> =
                when (effect) {
                    Effect.WRITING -> Writing(tupleSpaceName, before, tuple!!)
                    Effect.READING -> Reading(tupleSpaceName, before, tuple!!)
                    Effect.TAKING -> Taking(tupleSpaceName, before, tuple!!)
                    Effect.MISSING -> Missing(tupleSpaceName, before, template!!, tuple)
                }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> beforeWriting(tupleSpaceName: String, tuple: X): Writing<X, Y> {
            return Writing(tupleSpaceName, true, tuple)
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> afterWriting(tupleSpaceName: String, tuple: X): Writing<X, Y> {
            return Writing(tupleSpaceName, false, tuple)
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> beforeTaking(tupleSpaceName: String, tuple: X): Taking<X, Y> {
            return Taking(tupleSpaceName, false, tuple)
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> afterTaking(tupleSpaceName: String, tuple: X): Taking<X, Y> {
            return Taking(tupleSpaceName, true, tuple)
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> beforeReading(tupleSpaceName: String, tuple: X): Reading<X, Y> {
            return Reading(tupleSpaceName, false, tuple)
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> afterReading(tupleSpaceName: String, tuple: X): Reading<X, Y> {
            return Reading(tupleSpaceName, true, tuple)
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> beforeAbsent(tupleSpaceName: String, template: Y): Missing<X, Y> {
            return Missing(tupleSpaceName, false, template)
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> afterAbsent(tupleSpaceName: String, template: Y): Missing<X, Y> {
            return Missing(tupleSpaceName, true, template)
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> beforeAbsent(tupleSpaceName: String, template: Y, counterexample: X): Missing<X, Y> {
            return Missing(tupleSpaceName, false, template, counterexample)
        }

        @JvmStatic
        fun <X : Tuple<X>, Y : Template<X>> afterAbsent(tupleSpaceName: String, template: Y, counterexample: X): Missing<X, Y> {
            return Missing(tupleSpaceName, true, template, counterexample)
        }
    }
}